cognify/optimizer/core/sh.py

```diff
@@ -0,0 +1,121 @@
+from typing import Callable
+import logging
+from concurrent.futures import (
+    ThreadPoolExecutor,
+    as_completed,
+)
+from concurrent.futures import Future
+
+from cognify.optimizer.evaluator import (
+    EvaluationResult,
+    GeneralEvaluatorInterface,
+)
+from cognify.optimizer.core.common_stats import CommonStats
+from cognify.optimizer.core.flow import TopDownInformation
+
+logger = logging.getLogger(__name__)
+
+class SuccessiveHalving:
+    """Policy for successive halving resource allocation
+    
+    If a layer (A) adopts this policy, it will allocate resources for the next layer (B)
+    
+    layer A will first propose `n` trials, then
+    
+    **SH routine**:
+    1. each remaining trial at layer B will have a fixed step_budget = `r_i`
+    2. after all trials are evaluated, unpromising trials will be pruned
+    3. repeat from step 1 until no trials left or iteration limit reached
+    """
+    def __init__(
+        self,
+        prune_rate: int,
+        num_SH_iter: int,
+        initial_step_budget: int,
+        hierarchy_level: int,
+        next_layer_factory: Callable[[], GeneralEvaluatorInterface],
+        selected_runs: list[TopDownInformation],
+    ):
+        self.prune_rate = prune_rate
+        self.num_SH_iter = num_SH_iter
+        self.initial_step_budget = initial_step_budget
+        self.hierarchy_level = hierarchy_level
+        self.selected_runs: list[TopDownInformation] = selected_runs
+        self.ready_to_run = [i for i in range(len(selected_runs))]
+        self.num_inner_trials = [0] * len(selected_runs)
+        self._next_layer_factory = next_layer_factory
+    
+    def _evaluate_one_config(self, i):
+        _next_layer = self._next_layer_factory()
+        # NOTE: frac not needed bc next layer will never be evaluator
+        result = _next_layer.evaluate(self.selected_runs[i])
+        return result
+
+    def _halving_by_order(self, indicators, not_converged, n_left):
+        sorted_indicator_indices = sorted(
+            not_converged,
+            key=lambda x: (-indicators[x][0], indicators[x][2], indicators[x][1])
```

Comment:
What do the indices here mean? This seems error-prone, or we should be explicit about it

Reply:
this seems legacy code, removed now